### PR TITLE
contrib/gorm.io/gorm.v1: propagate parent span

### LINKED ISSUE #849: contrib/gorm.io/gorm.v1: child database spans are not attached to the gorm parent span
(As discussed on https://github.com/DataDog/dd-trace-go/pull/759)

If both gorm & the underlying `*sql.DB` connection are setup for tracing, spans for the underlying database connection queries are not attached to the gorm span created for the query.

Screenshot 2021-02-18 at 11 10 43

It's possible I've misconfigured something but it does seem like an actual issue from the above screenshot (the purple `postgres.query` span should be below the `gorm.query` span). I think this isn't an issue for `jinzhu/gorm` as there's no way to inject a traced `sql.DB` in the old API so you wouldn't have any sql spans at all.

For reproducing, we've got something like this setup:
```go
import (
  sqltrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
  gormtrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/gorm.io/gorm.v1"
)

sqltrace.Register("postgres", &pq.Driver{})
db, _ := sqltrace.Open("postgres", "postgres://")
orm, _ := gormtrace.Open(postgres.New(postgres.Config{
  Conn: db,
})
```

### PR DESCRIPTION
**Pull request by System-Glitch**
### What does this PR do?

- Add a way to use the gorm tracing as a gorm plugin instead of only having the `Open()` option. 
- Start span in before callback instead of after to allow child context to be given to underlying DB driver
- Fix row and raw callbacks not being ordered correctly
- Don't generate spans for `DryRun` statements (e.g. when using a statement or subquery as var)
- Improved tests for the gorm integration

fixes #849 

### Motivation

We are using this plugin as well as the postgres one together and were observing incorrect or confusing spans.

### Reviewer's Checklist

- [x] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

### CODE REVIEW

**Review by darccio — APPROVED** [review]
LGTM

File: contrib/gorm.io/gorm.v1/example_test.go [hunk 1]
```diff
@@ -41,6 +42,30 @@ func ExampleOpen() {
 	db.Where("name = ?", "jinzhu").First(&user)
 }
 
+func ExamplePlugin() {
```
Comment (rarguelloF) [inline contrib/gorm.io/gorm.v1/example_test.go:45]:
```suggestion
// ExampleNewTracePlugin illustrates how to trace gorm using the gorm.Plugin api. 
func ExampleNewTracePlugin() {
```

(so the example shows up with the `NewTracePlugin` documentation)

File: contrib/gorm.io/gorm.v1/gorm_test.go [hunk 2]
```diff
@@ -215,15 +216,26 @@ func TestCallbacks(t *testing.T) {
...
 		a.Equal(queryText, span.Tag(ext.ResourceName))
 		a.Equal("gorm.io/gorm.v1", span.Tag(ext.Component))
+		a.Equal(parentSpan.Context().SpanID(), span.ParentID())
+
+		for _, s := range spans {
+			if s.Tag(ext.Component) == "jackc/pgx.v5" {
+				// The underlying driver should receive the gorm span
+				a.Equal(span.SpanID(), s.ParentID())
+			}
+		}
```
Comment (rarguelloF) [inline contrib/gorm.io/gorm.v1/gorm_test.go:226]:
```suggestion
		spans := mt.FinishedSpans()
		require.Len(t, spans, 5)
		
		{
			parentSpan := spans[4]
			gormSpan := spans[3]
			pgx1 := spans[2]
			pgx2 := spans[1]
			pgx3 := spans[0]

			assert.Equal(t, "gorm.create", gormSpan.OperationName())
			assert.Equal(t, ext.SpanTypeSQL, gormSpan.Tag(ext.SpanType))
			assert.Equal(t, queryText, gormSpan.Tag(ext.ResourceName))
			assert.Equal(t, "gorm.io/gorm.v1", gormSpan.Tag(ext.Component))

			assert.Equal(t, gormSpan.ParentID(), parentSpan.SpanID())
			assert.Equal(t, pgx1.ParentID(), gormSpan.SpanID())
			assert.Equal(t, pgx2.ParentID(), gormSpan.SpanID())
			assert.Equal(t, pgx3.ParentID(), gormSpan.SpanID())
		}
```

I think I would prefer the test to be a little bit "dumber". Now that the `mocktracer.Start()` lines are added to each specific subtest, we can assert a concrete number of spans, and the expected behavior is more explicit.

(feel free to ignore this refactor if you prefer so, I can do it myself in a subsequent PR)

**Review by rarguelloF — CHANGES_REQUESTED** [review]
Thanks for the improvements and the fix 🎉 
I've requested some minor changes. Feel free to skip the test refactor, I can do that myself in a follow-up PR.

**Review by rarguelloF — APPROVED** [review]

**Pull request merged**